src/main/scala/mesosphere/marathon/core/task/tracker/impl/InstanceTrackerDelegate.scala

```diff
@@ -110,6 +128,18 @@ private[tracker] class InstanceTrackerDelegate(
 
         effectF // We already completed the sender promise with the future result (failed or not)
           .transform(_ => Success(Done)) // so here we map the future to a successful one to preserve the stream
+      case QueuedQuery(query, promise) =>
```

Comment:
I am still thinking about it. Basically what you did here is similar to if databases would acquire exclusive lock when writing OR reading which means that nobody would be able to read when someone is trying to write to that row. That just does not seem like a very performant solution :( and to be honest, InstanceTracker is kind of "database" in our Marathon world.

E.g. for TaskLauncherActor you really don't want to do this, I think that would have pretty big performance penalty.

Reply:
It kind of depends what guarantees we want to give. How can we model a fetch and guarantee that all updates which came in before have been processed? Or rather all updates that are on their way to the storage?

Reply:
yeah sure. it's the same in databases :) either you need exclusive lock or not. I think that considering current implementation of TaskReplaceActor, that guy needs this guarantee right now. But many other places using this method does not. E.g. I've seen scale check is using this as well, that for sure does not need that. So my proposal is just to use this exclusive lock only when absolutely necessary, so for now just in TaskReplaceActor

Reply:
Also, the [LazyCachingPersistenceStore](https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/core/storage/store/impl/cache/LazyCachingPersistenceStore.scala#L144) locks on `category` which resolve to simply `"instance"` in [ZkStoreSerialization](https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/storage/store/ZkStoreSerialization.scala#L66) and [InMemoryStoreSerialization](https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/storage/store/InMemoryStoreSerialization.scala#L47). So all updates for all instances are already sequentialized.

Reply:
yeah serializing on writes is normal. That's what SERIALIZABLE does on databases. But you can still read (old value) when write transaction is in process. So this is even more strict than that...

Reply:
Ah, true. The store locks on the storage id. Anyhow, what do you suggest we should do?

Reply:
@jeschkies my proposal would be what I wrote here https://github.com/mesosphere/marathon/pull/6893#pullrequestreview-225359836